File: examples/csharp/Product/src/ProductService.cs
using Angzarr;
using Examples;
using Grpc.Core;
using Serilog;

namespace Angzarr.Examples.Product;

public class ProductService : BusinessLogic.BusinessLogicBase
{
    private readonly IProductLogic _logic;
    private readonly Serilog.ILogger _logger;

    public ProductService(IProductLogic logic, Serilog.ILogger logger)
    {
        _logic = logic;
        _logger = logger.ForContext<ProductService>();
    }

    public override Task<BusinessResponse> Handle(ContextualCommand request, ServerCallContext context)
    {
        try
        {
            var events = ProcessCommand(request);
            var response = new BusinessResponse { Events = events };
            return Task.FromResult(response);
        }
        catch (CommandValidationException ex)
        {
            throw new RpcException(new Status(ex.StatusCode, ex.Message));
        }
        catch (Exception ex)
        {
            _logger.Error(ex, "Unexpected error processing command");
            throw new RpcException(new Status(StatusCode.Internal, $"Internal error: {ex.Message}"));
        }
    }

    private EventBook ProcessCommand(ContextualCommand request)
    {
        var cmdBook = request.Command;
        var priorEvents = request.Events;

        if (cmdBook == null || cmdBook.Pages.Count == 0)
            throw CommandValidationException.InvalidArgument("CommandBook has no pages");

        var cmdPage = cmdBook.Pages[0];
        if (cmdPage.Command == null)
            throw CommandValidationException.InvalidArgument("Command page has no command");

        var state = _logic.RebuildState(priorEvents);
        var command = cmdPage.Command;
        var typeUrl = command.TypeUrl;

        EventBook result;

        if (typeUrl.EndsWith("CreateProduct"))
        {
            var cmd = command.Unpack<CreateProduct>();
            result = _logic.HandleCreateProduct(state, cmd.Sku, cmd.Name, cmd.Description, cmd.PriceCents);
        }
        else if (typeUrl.EndsWith("UpdateProduct"))
        {
            var cmd = command.Unpack<UpdateProduct>();
            result = _logic.HandleUpdateProduct(state, cmd.Name, cmd.Description);
        }
        else if (typeUrl.EndsWith("SetPrice"))
        {
            var cmd = command.Unpack<SetPrice>();
            result = _logic.HandleSetPrice(state, cmd.PriceCents);
        }
        else if (typeUrl.EndsWith("Discontinue"))
        {
            result = _logic.HandleDiscontinue(state);
        }
        else
        {
            throw CommandValidationException.InvalidArgument($"Unknown command type: {typeUrl}");
        }

        result.Cover = cmdBook.Cover;
        return result;
    }
}

File: examples/cpp/product/src/product_service.cc
#include "product_logic.hpp"
#include "logging.hpp"
#include "validation_error.hpp"
#include "angzarr.grpc.pb.h"
#include <grpcpp/grpcpp.h>

namespace product {

class ProductService final : public angzarr::BusinessLogic::Service {
public:
    grpc::Status Handle(grpc::ServerContext* context,
                        const angzarr::ContextualCommand* request,
                        angzarr::BusinessResponse* response) override {
        try {
            const auto& cmd_book = request->command();
            const auto* prior_events = request->has_events() ? &request->events() : nullptr;

            if (cmd_book.pages().empty()) {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "CommandBook has no pages");
            }

            const auto& cmd_page = cmd_book.pages(0);
            if (!cmd_page.has_command()) {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Command page has no command");
            }

            auto state = ProductLogic::rebuild_state(prior_events);
            const auto& type_url = cmd_page.command().type_url();

            google::protobuf::Any event_any;

            if (type_url.find("CreateProduct") != std::string::npos) {
                examples::CreateProduct cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("product", "creating_product",
                    {{"sku", cmd.sku()}, {"name", cmd.name()}, {"price_cents", cmd.price_cents()}});
                auto event = ProductLogic::handle_create_product(
                    state, cmd.sku(), cmd.name(), cmd.description(), cmd.price_cents());
                event_any.PackFrom(event);
            } else if (type_url.find("UpdateProduct") != std::string::npos) {
                examples::UpdateProduct cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("product", "updating_product", {{"name", cmd.name()}});
                auto event = ProductLogic::handle_update_product(state, cmd.name(), cmd.description());
                event_any.PackFrom(event);
            } else if (type_url.find("SetPrice") != std::string::npos) {
                examples::SetPrice cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("product", "setting_price", {{"price_cents", cmd.price_cents()}});
                auto event = ProductLogic::handle_set_price(state, cmd.price_cents());
                event_any.PackFrom(event);
            } else if (type_url.find("Discontinue") != std::string::npos) {
                examples::Discontinue cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("product", "discontinuing_product", {{"reason", cmd.reason()}});
                auto event = ProductLogic::handle_discontinue(state, cmd.reason());
                event_any.PackFrom(event);
            } else {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Unknown command type: " + type_url);
            }

            auto* event_book = response->mutable_events();
            if (cmd_book.has_cover()) *event_book->mutable_cover() = cmd_book.cover();

            auto* event_page = event_book->add_pages();
            event_page->set_num(0);
            *event_page->mutable_event() = event_any;
            event_page->mutable_created_at()->set_seconds(
                std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));

            return grpc::Status::OK;
        } catch (const angzarr::ValidationError& e) {
            return e.to_grpc_status();
        }
    }
};

std::unique_ptr<angzarr::BusinessLogic::Service> create_product_service() {
    return std::make_unique<ProductService>();
}

}  // namespace product
